Ryujinx.Common/Collections/IntervalTree.cs

```diff
@@ -0,0 +1,820 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+
+namespace Ryujinx.Common.Collections
+{
+    /// <summary>
+    /// An Augmented Interval Tree based off of the "TreeDictionary"'s Red-Black Tree. Allows fast overlap checking of ranges.
+    /// </summary>
+    /// <typeparam name="K">Key</typeparam>
+    /// <typeparam name="V">Value</typeparam>
+    public class IntervalTree<K, V> where K : IComparable<K>
+    {
+        private const int ArrayGrowthSize = 32;
+
+        private const bool Black = true;
+        private const bool Red = false;
+        private IntervalTreeNode<K, V> _root = null;
+        private int _count = 0;
+
+        public int Count => _count;
+
+        public IntervalTree() { }
+
+        #region Public Methods
+
+        /// <summary>
+        /// Gets the values of the interval whose key is <paramref name="key"/>.
+        /// </summary>
+        /// <param name="key">Key of the node value to get</param>
+        /// <param name="overlaps">Overlaps array to place results in</param>
+        /// <returns>Number of values found</returns>
+        /// <exception cref="ArgumentNullException"><paramref name="key"/> is null</exception>
+        public int Get(K key, ref V[] overlaps)
+        {
+            if (key == null)
+            {
+                throw new ArgumentNullException(nameof(key));
+            }
+
+            IntervalTreeNode<K, V> node = GetNode(key);
+
+            if (node == null)
+            {
+                return 0;
+            }
+
+            if (node.Values.Count > overlaps.Length)
+            {
+                Array.Resize(ref overlaps, node.Values.Count);
+            }
+
+            int overlapsCount = 0;
+            foreach (RangeNode<K, V> value in node.Values)
+            {
+                overlaps[overlapsCount++] = value.Value;
+            }
```

Comment:
Perhaps something like this?

```suggestion
            int overlapsCount = node.Values.Count;
            
            if (overlapsCount > overlaps.Length)
            {
                overlaps = node.Values.ToArray();
            }
            else
            {
                node.Values.CopyTo(overlaps);
            }
```

Reply:
Can't do this - the `.Value` property must be extracted from the range node.

Reply:
Oof 😢